### PR TITLE
crimson/os/seastore: refactor transaction_manager and below to use interruptible_future

### PR DESCRIPTION
**Pull request by athanatos**
<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/crimson/common/errorator.h [hunk 1]
```diff
@@ -1211,6 +1211,18 @@ struct futurize<Container<::crimson::errorated_future_marker<Value>>> {
   using type = typename errorator_type::template future<Value>;
   using value_type = seastar::internal::future_stored_type_t<Value>;
 
+  template<typename Func, typename... FuncArgs>
+  [[gnu::always_inline]]
+    static inline type apply(Func&& func, std::tuple<FuncArgs...>&& args) noexcept {
```
Comment (tchaikov) [inline src/crimson/common/errorator.h:1216]: nit, might want to drop `inline` and reduce the indent by two spaces.

File: src/crimson/common/interruptible_future.h [hunk 2]
```diff
@@ -513,6 +523,10 @@ struct interruptible_errorator {
       InterruptCond, typename Errorator::template future<>>(
 	Errorator::now());
   }
+
+  using pass_further = typename Errorator::pass_further;
+  using ready_future_marker = ready_future_marker;
```
Comment (tchaikov) [inline src/crimson/common/interruptible_future.h]: just wanted to ensure that i understand the intention here: so you think it's more handy to have them in `interruptible_errorator`?

File: src/crimson/common/interruptible_future.h [hunk 3]
```diff
@@ -910,31 +951,51 @@ struct interruptor
...
+      std::forward<OnInterrupt>(efunc),
+      InterruptCond(std::forward<InterruptCondParams>(params)...));
+  }
+
+  template <typename Error,
+	    typename Func,
+	    typename... Params>
+  static inline auto with_interruption_to_error(
+    Func &&f, InterruptCond &&cond, Params&&... params) {
+    using func_result_t = decltype(f(std::forward<Params>(params)...));
```
Comment (tchaikov) [inline src/crimson/common/interruptible_future.h:980]: nit, would be more idiomatic to put:
```c++
using func_result_t = std::invoke_result_t<Func, Params...>;
```

File: src/crimson/common/interruptible_future.h (hunk 2 above)
Comment (athanatos) [reply src/crimson/common/interruptible_future.h]: Yep, this way it matches a normal errorator and I don't have to swap around the code.

**Comment by athanatos** [comment]
Hrm, I'm not sure why it's failing.  I don't see an error in the jenkins output except that the build was halted.

**Comment by cyx1231st** [comment]
jenkins test make check

**Comment by cyx1231st** [comment]
Looks linking has stalled:
```
[2448/2452] Linking CXX executable bin/unittest-object-data-handler
[2449/2452] Linking CXX executable bin/unittest-collection-manager
[2450/2452] Linking CXX executable bin/crimson-osd
```

**Comment by tchaikov** [comment]
```
USER         PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND
...
jenkins+ 3388173  100  0.2 600924 543380 ?       R    01:53  33:58 /usr/lib/gcc/x86_64-linux-gnu/9/cc1plus -quiet -I src/include -I ../src -I ../src/seastar/include -I src/seastar/gen/include -imultiarch x86_64-linux-gnu -D_GNU_SOURCE -D BOOST_ALL_NO_LIB -D BOOST_ASIO_DISABLE_THREAD_KEYWORD_EXTENSION -D BOOST_ASIO_USE_TS_EXECUTOR_AS_DEFAULT -D BOOST_PROGRAM_OPTIONS_DYN_LINK -D BOOST_THREAD_DYN_LINK -D HAVE_CONFIG_H -D SEASTAR_API_LEVEL=6 -D SEASTAR_DEBUG -D SEASTAR_DEBUG_SHARED_PTR -D SEASTAR_DEFAULT_ALLOCATOR -D SEASTAR_SHUFFLE_TASK_QUEUE -D SEASTAR_TYPE_ERASE_MORE -D WITH_SEASTAR=1 -D _FILE_OFFSET_BITS=64 -D _GNU_SOURCE -D _REENTRANT -D _THREAD_SAFE -D __CEPH__ -D __STDC_FORMAT_MACROS -D __linux__ -D _GLIBCXX_ASSERTIONS -U _FORTIFY_SOURCE -D SEASTAR_SSTRING -isystem /opt/ceph/include -isystem include -isystem ../src/xxHash -isystem ../src/rapidjson/include -isystem ../src/googletest/googlemock/include -isystem ../src/googletest/googlemock -isystem ../src/googletest/googletest/include -isystem ../src/googletest/googletest ../src/test/crimson/test_interruptible_future.cc -quiet -dumpbase test_interruptible_future.cc -mtune=generic -march=x86-64 -auxbase-strip src/test/crimson/CMakeFiles/unittest-interruptible-future.dir/test_interruptible_future.cc.o -g -Wall -Wtype-limits -Wpointer-arith -Werror=format-security -Winit-self -Wno-unknown-pragmas -Wno-ignored-qualifiers -Wpessimizing-move -Wredundant-move -Wstrict-null-sentinel -Woverloaded-virtual -Wno-non-virtual-dtor -Wno-maybe-uninitialized -Wno-error=unused-result -std=gnu++17 -std=c++17 -fdiagnostics-color=auto -fPIE -fno-strict-aliasing -fsigned-char -ftemplate-depth=1024 -fno-new-ttp-matching -fstack-protector-strong -fno-builtin-malloc -fno-builtin-calloc -fno-builtin-realloc -fno-builtin-free -fstack-clash-protection -fsanitize=address -fsanitize=undefined -fno-sanitize=vptr -fasynchronous-unwind-tables -fstack-protector-strong -Wformat-security -fstack-clash-protection -fcf-protection -o /tmp/ccjkDBeK.s
```

**Comment by athanatos** [comment]
I'll get a bionic env together tomorrow and try to reproduce.

File: src/crimson/common/interruptible_future.h [hunk 4]
```diff
@@ -1181,6 +1181,11 @@ struct interruptor
     }
   }
 
+  template <typename... Args>
+  static auto do_until(Args&&... args) {
```
Comment (tchaikov) [inline src/crimson/common/interruptible_future.h:1133]: created #41984 to address this issue.

**Review by tchaikov — APPROVED** [review]

**Comment by athanatos** [comment]
@tchaikov rebased, removed the do_until patch.

**Comment by liu-chunmei** [comment]
I am wondering why we don't we use seastar::mutex to avoid extent confliction.  once extent is invalid or retired, the transaction should begin from the beginning,  why don't we add a seastar::mutex to each cached extent? I think that will be more efficient than repeat and check interruption.

**Comment by athanatos** [comment]
Every single transaction would lock the root node.  Using read/write mutexes as an optimization requires predicting prior to descent which node will be the first to be mutated.  We may need to add some kind of heuristic pessimistic concurrency control later on, but using this form of optimistic concurrency control should permit more concurrent transactions in the common case.

File: src/test/crimson/test_interruptible_future.cc [hunk 5]
```diff
@@ -148,20 +150,97 @@ TEST_F(seastar_test_suite_t, loops)
...
+	  1
+	).si_then([](auto) {
+	  return base2_iertr::make_ready_future<>();
+	});
+      });
+    ret.unsafe_get0();
+  });
+}
+
+TEST_F(seastar_test_suite_t, handle_error)
```
Comment (tchaikov) [inline src/test/crimson/test_interruptible_future.cc:230]: this is the test which can keep GCC-9 busy forever.

Comment (athanatos) [reply src/test/crimson/test_interruptible_future.cc:230]: Interesting.  Thank!  I think I'll commit it out for now?

**Comment by athanatos** [comment]
Test failures, working on it.

**Comment by liu-chunmei** [comment]
@athanatos I download the current master and Debug build, got the following assert.

crimson-osd: ../src/crimson/os/seastore/cache.h:184: crimson::os::seastore::Cache::get_extent(crimson::os::seastore::paddr_t, crimson::os::seastore::segment_off_t) [with T = crimson::os::seastore::collection_manager::CollectionNode; crimson::os::seastore::Cache::get_extent_ret<T> = crimson::errorator<crimson::unthrowable_wrapper<const std::error_code&, ((const std::error_code&)(& crimson::ec<std::errc::io_error>))> >::_future<crimson::errorated_future_marker<boost::intrusive_ptr<crimson::os::seastore::collection_manager::CollectionNode> > >; crimson::os::seastore::segment_off_t = int]::<lambda()> mutable: Assertion `ret->is_valid()' failed.
Aborting on shard 0.

**Comment by liu-chunmei** [comment]
@athanatos Does Onode tree and Omap tree need do something for using interruptible future?

**Comment by athanatos** [comment]
Yeah, just ran into that myself.  About to push a PR.

**Comment by athanatos** [comment]
@liu-chunmei https://github.com/ceph/ceph/pull/42018

**Comment by athanatos** [comment]
@liu-chunmei   The next step will be for onode tree, omap tree, etc to convert to using interruptible_futures as well.

**Pull request merged**